asserts/model.go

```diff
@@ -25,16 +25,310 @@ import (
+	addSnap := func(snap *ModelSnap, bootSnap int) {
+		if snap == nil {
+			return
+		}
+		numBootSnaps += bootSnap
+		allSnaps = append(allSnaps, snap)
+		if snap.Presence == "required" {
+			allRequiredSnaps = append(allRequiredSnaps, snap)
+		}
+	}
+
+	addSnap(ms.Base, 1)
+	addSnap(ms.Gadget, 1)
+	addSnap(ms.Kernel, 1)
+	for _, snap := range ms.Snaps {
+		addSnap(snap, 0)
+	}
+	return allSnaps, allRequiredSnaps, numBootSnaps
+}
+
+var (
+	bootSnapModes = []string{"run", "ephemeral"}
+	defaultModes  = []string{"run"}
+)
+
+func checkExtendSnaps(extendedSnaps interface{}, base string) (*modelSnaps, error) {
+	const wrongHeaderType = `"snaps" header must be a list of maps`
+
+	entries, ok := extendedSnaps.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf(wrongHeaderType)
+	}
+
+	var modelSnaps modelSnaps
+	seen := make(map[string]bool, len(entries))
+	seenIDs := make(map[string]string, len(entries))
+
+	for _, entry := range entries {
+		snap, ok := entry.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf(wrongHeaderType)
+		}
+		modelSnap, err := checkModelSnap(snap)
+		if err != nil {
+			return nil, err
+		}
+
+		if seen[modelSnap.Name] {
+			return nil, fmt.Errorf("cannot list the same snap %q multiple times", modelSnap.Name)
+		}
+		if underName := seenIDs[modelSnap.SnapID]; underName != "" {
+			return nil, fmt.Errorf("cannot specify the same snap id %q multiple times, specified for snaps %q and %q", modelSnap.SnapID, underName, modelSnap.Name)
+		}
+		seen[modelSnap.Name] = true
+		seenIDs[modelSnap.SnapID] = modelSnap.Name
+
+		boot := false
+		switch {
+		case modelSnap.SnapType == "kernel":
+			boot = true
+			if modelSnaps.Kernel != nil {
+				return nil, fmt.Errorf("cannot specify multiple kernel snaps: %q and %q", modelSnaps.Kernel.Name, modelSnap.Name)
+			}
+			modelSnaps.Kernel = modelSnap
+		case modelSnap.SnapType == "gadget":
+			boot = true
+			if modelSnaps.Gadget != nil {
+				return nil, fmt.Errorf("cannot specify multiple gadget snaps: %q and %q", modelSnaps.Gadget.Name, modelSnap.Name)
+			}
+			modelSnaps.Gadget = modelSnap
+		case modelSnap.Name == base:
+			boot = true
+			if modelSnap.SnapType != "base" {
+				return nil, fmt.Errorf(`boot base %q must specify type "base", not %q`, base, modelSnap.SnapType)
+			}
+			modelSnaps.Base = modelSnap
+		}
+
+		if boot {
+			if len(modelSnap.Modes) != 0 || modelSnap.Presence != "" {
+				return nil, fmt.Errorf("boot snaps are always available, cannot specify modes or presence for snap %q", modelSnap.Name)
+			}
+			modelSnap.Modes = bootSnapModes
+		}
+
+		if len(modelSnap.Modes) == 0 {
+			modelSnap.Modes = defaultModes
+		}
+		if modelSnap.Presence == "" {
+			modelSnap.Presence = "required"
+		}
+
+		if !boot {
+			modelSnaps.Snaps = append(modelSnaps.Snaps, modelSnap)
+		}
+	}
+
+	return &modelSnaps, nil
+}
+
+var (
+	validSnapTypes     = []string{"app", "base", "gadget", "kernel", "core"}
```

Comment:
I know, it's complicated.  It's "core" and not "os", because that's the spelling in assertions of the type, so we'll have to translate somewhere, it's confusing, but is already like that in snap-declaration. About type snapd,  in principle it is implict, but people might want to fix a track for it, but I want to add it later for now.

Reply:
Anyway I thought of having a snap/concepts or snap/types to avoid this but feels like a follow up to me when some other things are clearer.

Reply:
Sounds good, thanks for the explanation!